src/bors/mergeable_queue.rs

```diff
@@ -0,0 +1,316 @@
+use crate::database::OctocrabMergeableState;
+use crate::github::{GithubRepoName, PullRequestNumber};
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+use tokio::sync::Notify;
+use tokio::time::timeout;
+
+use super::BorsContext;
+
+/// Delay before processing a mergeable queue item for the first time.
+const BASE_DELAY: Duration = Duration::from_millis(500);
+/// Exponential backoff delay multiplier.
+const BACKOFF_MULTIPLIER: f64 = 2.0;
+/// Max number of mergeable check retries before giving up.
+const MAX_RETRIES: u32 = 5;
+
+#[derive(Debug, Clone)]
+pub struct QueuedPullRequest {
+    pub pr_number: PullRequestNumber,
+    pub repo: GithubRepoName,
+}
+
+impl std::fmt::Display for QueuedPullRequest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}/#{}", self.repo, self.pr_number)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct MergeableQueueItem {
+    pub pull_request: QueuedPullRequest,
+    pub attempt: u32,
+    pub immediate: bool,
+}
+
+#[derive(Debug, Clone)]
+struct Item {
+    /// When to process item (None = immediate).
+    /// Reversed to create min-heap for expirations.
+    expiration: Reverse<Option<Instant>>,
+    inner: MergeableQueueItem,
+}
+
+impl PartialEq for Item {
+    fn eq(&self, other: &Self) -> bool {
+        self.expiration == other.expiration
+    }
+}
+
+impl Eq for Item {}
+
+impl PartialOrd for Item {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Item {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.expiration.cmp(&other.expiration)
+    }
+}
+
+struct SharedInner {
+    queue: Mutex<BinaryHeap<Item>>,
+    notify: Notify,
+    closed: AtomicBool,
+}
+
+#[derive(Clone)]
+pub struct MergeableQueueSender {
+    inner: Arc<SharedInner>,
+}
+
+#[derive(Clone)]
+pub struct MergeableQueueReceiver {
+    inner: Arc<SharedInner>,
+}
+
+pub fn create_mergeable_queue() -> (MergeableQueueSender, MergeableQueueReceiver) {
+    let shared = Arc::new(SharedInner {
+        queue: Mutex::new(BinaryHeap::new()),
+        notify: Notify::new(),
+        closed: AtomicBool::new(false),
+    });
+
+    (
+        MergeableQueueSender {
+            inner: shared.clone(),
+        },
+        MergeableQueueReceiver { inner: shared },
+    )
+}
+
+impl MergeableQueueSender {
+    pub fn close(&self) {
+        self.inner.closed.store(true, Ordering::Relaxed);
```

Comment:
Not sure if relaxed is enough, the compiler or the CPU could theoretically reorder this to be execute after the notify below. In theory we should use `acquire + release` semantics, but I don't want to think about that, so let's just use `SeqCst` everywhere.